File: python/problem28.py
class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        lenHaystack = len(haystack);
        lenNeedle = len(needle);
        if lenHaystack < lenNeedle:
            return -1;
        if lenHaystack == lenNeedle:
            if haystack == needle:
                return 0;
            else:
                return -1;
        if lenNeedle == 0:
            return 0;
        for i in range(lenHaystack - lenNeedle + 1):
            for j in range(lenNeedle):
                if haystack[i] != needle[j]:    #只要有一个字符不匹配就进行下一个匹配
                    break;
                else:
                    if j == lenNeedle - 1:      #匹配到最后一个字符说明匹配成功
                        return i - j;
                    i += 1;
        return -1;

File: Java/problem28.java
public class Solution 
{
    public int strStr(String haystack, String needle) 
    {
        //使用Boyer-Moore算法
        int lenHaystack = haystack.length();
        int lenNeedle = needle.length();
        if(lenNeedle == 0)
            return 0;
        int i = lenNeedle - 1, j = lenNeedle - 1;
        HashMap<Character, Integer> last = last_occur(needle);  //预处理获得各个字母在模式串中最后出现位置
        while(i < lenHaystack)
        {
            if(needle.charAt(j) == haystack.charAt(i))
            {
                if(j == 0)                      //匹配成功
                    return i;
                else                            //只是部分匹配成功，比较前一个字符
                {
                    --j;
                    --i;
                }
            }
            else                                //暂时失配，取下一个比较位置
            {
                i += lenNeedle - Math.min(j, last.get(haystack.charAt(i)) + 1);
                j = lenNeedle - 1;
            }
        }
        return -1;
    }
    private HashMap<Character, Integer> last_occur(String needle)
    {
        HashMap <Character, Integer> last = new HashMap<Character, Integer>();
        for(int i = 0; i < 26; ++i)
            last.put((char)('a' + i), -1);
        for(int i = 0; i < needle.length(); ++i)
            last.put(needle.charAt(i), i);
        return last;
    }
}